samples/subsys/usb/hid-cdc/src/main.c

```diff
@@ -0,0 +1,868 @@
+{
+	int ret;
+	struct app_evt_t *ev;
+	struct k_mem_block block;
+
+	ret = k_mem_pool_alloc(&event_elem_pool, &block,
+			       sizeof(struct app_evt_t),
+			       K_NO_WAIT);
+
+	if (ret < 0) {
+		LOG_ERR("APP event allocation failed!");
+		app_evt_flush();
+
+		ret = k_mem_pool_alloc(&event_elem_pool, &block,
+					       sizeof(struct app_evt_t),
+					       K_NO_WAIT);
+		if (ret < 0) {
+			LOG_ERR("APP event memory corrupted.");
+			__ASSERT_NO_MSG(0);
+			return NULL;
+		}
+		return NULL;
+	}
+
+	ev = (struct app_evt_t *)block.data;
+	ev->block = block;
+
+	return ev;
+}
+
+/* HID */
+
+static const u8_t hid_mouse_report_desc[] = HID_MOUSE_REPORT_DESC(2);
+static const u8_t hid_kbd_report_desc[] = HID_KEYBOARD_REPORT_DESC();
+
+static K_SEM_DEFINE(evt_sem, 0, 1);		/* starts off "not available" */
+static K_SEM_DEFINE(usb_sem, 1, 1);	/* starts off "available" */
+static struct gpio_callback callback[4];
+
+static char data_buf_mouse[64], data_buf_kbd[64];
+static char string[64];
+static u8_t chr_ptr_mouse, chr_ptr_kbd, str_pointer;
+bool caps;
+
+#define MOUSE_BTN_REPORT_POS	0
+#define MOUSE_X_REPORT_POS	1
+#define MOUSE_Y_REPORT_POS	2
+
+#define MOUSE_BTN_LEFT		BIT(0)
+#define MOUSE_BTN_RIGHT		BIT(1)
+#define MOUSE_BTN_MIDDLE	BIT(2)
+
+static const char *banner0  =	"Welcome to CDC ACM 0!\r\n"
+				"Supported commands:\r\n"
+				"up    - moves the mouse up\r\n"
+				"down  - moves the mouse down\r\n"
+				"right - moves the mouse to right\r\n"
+				"left  - moves the mouse to left\r\n";
+static const char *banner1  =	"Welcome to CDC ACM 1!\r\n"
+				"Enter a string and terminate "
+				"it with ENTER.\r\n"
+				"It will be sent via HID "
+				"when BUTTON 2 is pressed.\r\n"
+				"You can modify it by sending "
+				"a new one here.\r\n";
+static const char *gpio0    =	"Button 0 pressed\r\n";
+static const char *gpio1    =	"Button 1 pressed\r\n";
+static const char *gpio2    =	"Button 2 pressed\r\n";
+static const char *gpio3    =	"Button 3 pressed\r\n";
+static const char *unknown  =	"Command not recognized.\r\n";
+static const char *up	    =	"Mouse up\r\n";
+static const char *down	    =	"Mouse down\r\n";
+static const char *left	    =	"Mouse left\r\n";
+static const char *right    =	"Mouse right\r\n";
+static const char *evt_fail =	"Unknown event detected!\r\n";
+static const char *set_str  =	"String set to: ";
+static const char *endl     =	"\r\n";
+
+static void in_ready_cb(void)
+{
+	k_sem_give(&usb_sem);
+}
+
+static const struct hid_ops ops = {
+	.int_in_ready = in_ready_cb,
+};
+
+static inline void clear_mouse_report(void)
+{
+	struct app_evt_t *new_evt = app_evt_alloc();
+
+	new_evt->event_type = HID_MOUSE_CLEAR;
+	app_evt_put(new_evt);
+	k_sem_give(&evt_sem);
+}
+
+static inline void clear_kbd_report(void)
+{
+	struct app_evt_t *new_evt = app_evt_alloc();
+
+	new_evt->event_type = HID_KBD_CLEAR;
+	app_evt_put(new_evt);
+	k_sem_give(&evt_sem);
+}
+
+static inline int ascii_to_hid(u8_t ascii)
```

Comment:
Why would you want to inline a 100 lines long `switch` case of application code?

Reply:
This function is called only from one place.